Fixes in locomotor path cache

teinarss opened the pull request:
Fixes #16845 
Fixes #16848
Fixes #16863
Fixes: #16862
Fixes: #16880

Note: Is not dependent on the suggested change on setting  MovementType to none after the movement is finished. But if we fix that we can cache the case when a crushable unit has moved.

teinarss committed:
Change impassable terrain to short.max

tovl commented on OpenRA.Mods.Common/Traits/World/Locomotor.cs:
```diff
@@ -292,7 +293,8 @@ bool CanMoveFreelyInto(Actor actor, Actor ignoreActor, CPos cell, CellConditions
 
-			if (cellBlocking.HasCellBlocking(CellBlocking.Crushable))
+            // We can crush this actor and it is not moving
+			if (cellBlocking.HasCellBlocking(CellBlocking.Crushable) && !cellBlocking.HasCellBlocking(CellBlocking.IsMoving))
```
Why would it matter if the blocking actor is moving when we can crush it anyway?

tovl commented on OpenRA.Mods.Common/Traits/World/Locomotor.cs:
```diff
@@ -461,12 +463,15 @@ void UpdateCellBlocking(CPos cell)
+
+					if (isMoving)
+					{
+						actorBits = actorBits.Except(actor.Owner.AllyMask);
```
We can only ignore moving allies when `!check.HasCellCondition(CellConditions.BlockedByMovers)`

tovl commented on OpenRA.Mods.Common/Traits/World/Locomotor.cs:
```diff
@@ -461,12 +463,15 @@ void UpdateCellBlocking(CPos cell)
+					if (isMoving)
+					{
+						actorBits = actorBits.Except(actor.Owner.AllyMask);
+						cellBlocking |= CellBlocking.IsMoving;
```
When _any_ actor in the cell is _not_ moving `IsMoving` should not be flagged. Might be easier to just reverse the meaning.

abcdefg30 commented on OpenRA.Mods.Common/Traits/World/Locomotor.cs:
```diff
@@ -292,7 +293,8 @@ bool CanMoveFreelyInto(Actor actor, Actor ignoreActor, CPos cell, CellConditions
 				return false;
 
-			if (cellBlocking.HasCellBlocking(CellBlocking.Crushable))
+            // We can crush this actor and it is not moving
```
Spaces instead of tabs here.

fusion809 reviewed (dismissed):
This PR seems to fix the problem (saying "seems to" as it is intermittent, so it is difficult to know for certain).

teinarss committed:
Fixes on blocking logic

teinarss commented:
Updated
Had to remove the cached crushable logic because Crushable trait is conditional
Added support to cache the logic to determine if the cell has a ITemporaryBlocker

pchote commented:
Conditional `Crushable` shouldn't be a problem, surely? We just need to (a) filter disabled traits when calculating the cached value and (b) make `Crushable` ping the `ActorMap` &rarr; `Locomotor` so its cells can be marked dirty and recalculated when it is enabled/disabled.

teinarss commented:
Would be better to do this in a anther PR, where we can remove the ICrushable interface altogether.

teinarss commented:
Updated, fixed the crushable thing now. Still have do the update when the condition changes.

pchote commented on OpenRA.Mods.Common/Traits/World/Locomotor.cs:
```diff
@@ -445,32 +455,52 @@ void UpdateCellBlocking(CPos cell)
+					{
+						// If there is a temporary blocker in this cell.
+						if (actor.TraitOrDefault<ITemporaryBlocker>() == null)
+							cellHasTemporaryBlocker = false;
```
Shouldn't this be false only if *all* actors in the cell don't have the interface?

abcdefg30 commented on OpenRA.Mods.Cnc/Traits/Mine.cs:
```diff
@@ -59,11 +59,15 @@ bool ICrushable.CrushableBy(Actor self, Actor crusher, BitSet<CrushClass> crushC
-			return false;
+			var blocking = default(LongBitSet<PlayerBitMask>);
+			if (info.BlockFriendly)
+				blocking = self.Owner.AllyMask;
```
Imho `var blocking = info.BlockingFriendly ? self.Owner.AllyMask : default(LongBitSet<PlayerBitMask>);` is a bit better.

teinarss replied:
Done!

abcdefg30 reviewed (dismissed):
Gates and transports work again. :white_check_mark:

ubitux commented:
Since you mentioned it was supposed to fix it, I tried your branch but can still reproduce this:

![stack-teinarss](https://user-images.githubusercontent.com/34467/62454558-d5a59a00-b774-11e9-902c-ef644d0d856c.gif)

abcdefg30 commented:
I suppose this can/needs to be fixed by another PR (since this PR doesn't make things worse).

teinarss committed:
Update locomotor cache when Crushable is enabled/disabled

teinarss commented:
Updated: hoping that everything will work now. Added a hashset containing all cells that have been updated but not rechecked, so when queried for the blocking it will look at that flag and determine if we need to update this cell with new info about blocking.

teinarss committed:
Reset Locomotor cache on world loaded

teinarss commented:
Added a fix on https://github.com/OpenRA/OpenRA/issues/14104#issuecomment-518256217

tovl commented on OpenRA.Mods.Common/Traits/World/Locomotor.cs:
```diff
@@ -288,31 +290,37 @@ bool CanMoveFreelyInto(Actor actor, Actor ignoreActor, CPos cell, CellConditions
 
-			if (cellBlocking.HasCellBlocking(CellBlocking.Crushable))
-				return true;
+			if (check == CellConditions.BlockedByMovers && cellBlocking < CellBlocking.Crushable)
```
Shouldn't this also check for `ignoreActor == null` ?

pchote commented on OpenRA.Mods.Cnc/Traits/Mine.cs:
```diff
@@ -59,11 +59,13 @@ bool ICrushable.CrushableBy(Actor self, Actor crusher, BitSet<CrushClass> crushC
-			return false;
+			var blocking = info.BlockFriendly ? self.Owner.AllyMask : default(LongBitSet<PlayerBitMask>);
+
+			blocking = blocking.Union(info.CrushClasses.Overlaps(crushClasses) ? default(LongBitSet<PlayerBitMask>) : self.World.AllPlayerMask);
```
Am I right in assuming this can be simplified down to
```csharp
if (!info.CrushClasses.Overlaps(crushClasses))
	return self.World.AllPlayerMask;

// Friendly units should move around!
return info.BlockFriendly ? self.Owner.AllyMask : default(LongBitSet<PlayerBitMask>);
```
?

pchote replied:
What do you think about introducing a `self.World.NoPlayersMask` (and renaming `AllPlayersMask`, `AlliedPlayersMask`, `EnemyPlayersMask` for consistency)?

IMO this would make code like this a lot more intuitive.

pchote commented on OpenRA.Mods.Common/Traits/Crushable.cs:
```diff
@@ -65,6 +65,14 @@ bool ICrushable.CrushableBy(Actor self, Actor crusher, BitSet<CrushClass> crushC
+		{
+			if (IsTraitDisabled || !self.IsAtGroundLevel() || !Info.CrushClasses.Overlaps(crushClasses))
+				return self.World.AllPlayerMask;
+			else
```
The `else` here is redundant: can remove it and unindent the `return` below.

pchote commented on OpenRA.Mods.Common/Traits/Crushable.cs:
```diff
@@ -80,14 +88,14 @@ bool CrushableInner(BitSet<CrushClass> crushClasses, Player crushOwner)
-				blocking = self.World.AllPlayerMask;
-			else
-				blocking = Info.CrushedByFriendlies ? default(LongBitSet<PlayerBitMask>) : self.Owner.AllyMask;
+			self.World.ActorMap.MarkAsDirty(self.OccupiesSpace);
```
Can we achieve the same thing by calling `ActorMap.UpdatePosition(self.OccupiesSpace)` instead?

pchote commented on OpenRA.Mods.Common/Traits/Mobile.cs:
```diff
@@ -71,6 +71,8 @@ public override void RulesetLoaded(Ruleset rules, ActorInfo ai)
 			else if (locomotorInfos.Count(li => li.Name == Locomotor) > 1)
 				throw new YamlException("There is more than one locomotor named '{0}'.".F(Locomotor));
 
+			locomotor = null;
```
Can we please have a comment here explaining why this is needed? At first look this seems to be redundant, but your comments in IRC suggest this is fixing a regression - I guess we are leaking state between worlds with the `ModData.DefaultRules`?

pchote commented on OpenRA.Mods.Common/Traits/World/ActorMap.cs:
```diff
@@ -173,7 +173,7 @@ public void Dispose()
 		readonly Dictionary<int, CellLayer<InfluenceNode>> customInfluence = new Dictionary<int, CellLayer<InfluenceNode>>();
 		public readonly Dictionary<int, ICustomMovementLayer> CustomMovementLayers = new Dictionary<int, ICustomMovementLayer>();
-		public event Action<IEnumerable<CPos>> CellsUpdated;
+		public event Action<CPos> UpdateCell;
```
Shouldn't this be `CellUpdated` for consistency?

pchote commented on OpenRA.Mods.Common/Traits/World/Locomotor.cs:
```diff
@@ -288,31 +290,37 @@ bool CanMoveFreelyInto(Actor actor, Actor ignoreActor, CPos cell, CellConditions
 
-			if (sharesCell && cellBlocking.HasCellBlocking(CellBlocking.FreeSubCell))
+			// We can only do an early exit here when we know the unit is not moving.
+			// Because we cant distinguish between crushable and ally moving (which we need to fall back to the old code)
```
IIRC there was some discussion earlier about inverting and replacing `IsMoving` with something like `StationaryBlocker`. IMO this would make the logic clearer for cells with a mix of stationary and moving units. 

`StationaryBlocker` retains its meaning if another unit in the cell is moving, whereas the current code will (i think?) set `IsMoving` if the cell has one moving unit, but clear it if there is one moving and one stationary unit.

pchote commented on OpenRA.Mods.Common/Traits/World/Locomotor.cs:
```diff
@@ -34,7 +34,9 @@ public enum CellBlocking : byte
-		Crushable = 4
+		IsMoving = 4,
+		Crushable = 8,
+		IsTemporaryBlocker = 16
```
How about `HasTemporaryBlocker`? `Is` implies that there is only one thing in the cell, which may not be true.

fusion809 commented:
I believe this PR has taken a step backward, as I've had a frozen engineer and a frozen ore truck on another team. I think this replay is the one that shows this (don't know, as it crashes before I can see if it is the right one):

[OpenRA-2019-08-06T121207Z.orarep.zip](https://github.com/OpenRA/OpenRA/files/3472430/OpenRA-2019-08-06T121207Z.orarep.zip)

my logs are here:

[OpenRA-2019-08-06.zip](https://github.com/OpenRA/OpenRA/files/3472435/OpenRA-2019-08-06.zip)

fusion809 commented:
Here's the exception log for the replay crash: [exception-2019-08-06T131527Z.log.zip](https://github.com/OpenRA/OpenRA/files/3472819/exception-2019-08-06T131527Z.log.zip).

matjaeck commented:
I didn't encounter any issues while testing and the bugs seem fixed, replays seem to work, too.

fusion809 commented:
I haven't been able to recreate it again yet, so maybe it was a fluke. Feel free to ignore, if it reappears that'd be for another bug report (assuming it's merged by then).

teinarss commented:
Updated

pchote commented on OpenRA.Game/Player.cs:
```diff
@@ -76,7 +76,7 @@ struct StanceColors
 		public LongBitSet<PlayerBitMask> PlayerMask;
-		public LongBitSet<PlayerBitMask> AllyMask = default(LongBitSet<PlayerBitMask>);
+		public LongBitSet<PlayerBitMask> AlliedPlayerMask = default(LongBitSet<PlayerBitMask>);
 		public LongBitSet<PlayerBitMask> EnemyMask = default(LongBitSet<PlayerBitMask>);
```
This will need to be updated to `EnemyPlayersMask` (can can we please also have `AlliedPlayersMask` and `AllPlayersMask` with `s`es?)

pchote commented on OpenRA.Mods.Cnc/Traits/Mine.cs:
```diff
@@ -59,11 +59,13 @@ bool ICrushable.CrushableBy(Actor self, Actor crusher, BitSet<CrushClass> crushC
-			// Fall back to the slow path
-			blocking = default(LongBitSet<PlayerBitMask>);
-			return false;
+			var blocking = info.BlockFriendly ? self.Owner.AlliedPlayerMask : self.World.NoPlayersMask;
```
Can we please simplify this down to

```csharp
if (!info.CrushClasses.Overlaps(crushClasses))
	return self.World.AllPlayersMask;

// Friendly units should move around!
return info.BlockFriendly ? self.Owner.AlliedPlayersMask : self.World.NoPlayersMask;
```
(assuming this simplification is correct)?

pchote commented on OpenRA.Mods.Common/Traits/Crates/Crate.cs:
```diff
@@ -225,14 +225,9 @@ bool ICrushable.CrushableBy(Actor self, Actor crusher, BitSet<CrushClass> crushC
-				blocking = self.World.AllPlayerMask;
-
-			return true;
+			return self.IsAtGroundLevel() && crushClasses.Contains(info.CrushClass) ? default(LongBitSet<PlayerBitMask>) : self.World.AllPlayerMask;
```
`default(LongBitSet<PlayerBitMask>)` &rarr; `self.World.NoPlayersMask`

pchote commented on OpenRA.Mods.Common/Traits/World/Locomotor.cs:
```diff
@@ -288,31 +290,37 @@ bool CanMoveFreelyInto(Actor actor, Actor ignoreActor, CPos cell, CellConditions
 
-			if (sharesCell && cellBlocking.HasCellBlocking(CellBlocking.FreeSubCell))
+			// We can only do an early exit here when we know the unit is not moving.
+			// Because we cant distinguish between crushable and ally moving (which we need to fall back to the old code)
```
Maybe I'm still not understanding the logic here, but shouldn't this be something like
```csharp
// Moving actors have direction-dependent blocking, which must be evaluated on the slow path
// unless we know we can crush the actor (crushing is direction independent)
if (!blocking.Overlaps(actor.Owner.PlayerMask) && (!cellBlocking.HasCellBlocking(CellBlocking.HasMovingActor) || cellBlocking.HasCellBlocking(CellBlocking.Crushable))
	return true;
```

If I'm indeed not understanding the logic that makes it even more important to have a clear comment!

teinarss commented:
Updated

pchote commented on OpenRA.Mods.Common/Traits/World/Locomotor.cs:
```diff
@@ -418,66 +417,72 @@ void UpdateCellCost(CPos cell)
-
-						if (isMoving)
-							actorBits = actorBits.Except(actor.Owner.AllyMask);
+                        cellFlag |= CellFlag.HasCrushableActor;
```
Space indentation needs to be replaced with tabs here

teinarss commented:
Updatet

pchote reviewed (commented):
LGTM, and thanks - the code is now much easier to understand.
Just a few more style / renaming requests to help the readability a bit further

pchote commented:
I had originally suggested that this fixes #16901, but it seems that this is only because the PR is based on an old bleed commit, before that regression was introduced. The problem returns when this is rebased on latest bleed.

teinarss commented:
updated

reaperrr commented on OpenRA.Mods.Common/Traits/Mobile.cs:
```diff
@@ -71,6 +71,9 @@ public override void RulesetLoaded(Ruleset rules, ActorInfo ai)
 			else if (locomotorInfos.Count(li => li.Name == Locomotor) > 1)
 				throw new YamlException("There is more than one locomotor named '{0}'.".F(Locomotor));
 
+			// We need to reset the reset the reference to the locomotor between each worlds, otherwise we are reference the previous state.
```
Double 'reset the'

reaperrr commented:
Didn't notice any issues while watching the RA shellmap for a while. LGTM after that typo fixup.

teinarss commented:
Updated!

reaperrr commented:
https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/9fab6c68b8e06af765c2663c81f5f6906f1afa17...6ac1647a39b8dbf71d3e37891c42030971cd4ad2
